### PR TITLE
IGNITE-19096 Sql. Replace placeholder values from ModifyNode

### PR DESCRIPTION
**Pull request by zstan**
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java [hunk 1]
```diff
@@ -69,6 +83,98 @@ public IgniteSqlToRelConvertor(
...
+
+            List<Pair<RexNode, String>> exps = new ArrayList<>(targetFields.size());
+
+            int pos = 0;
+            int rowPos = 0;
+            for (RelDataTypeField fld : tblFields) {
+                if (fld.getName().equals(targetFields.get(rowPos))) {
+                    SqlNode operand = rowConstructor0.getOperandList().get(rowPos);
+
+                    if (operand.getKind() == SqlKind.DEFAULT) {
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java:153]: Can operand be null?
Is 3-rd value allowed:  INSERT INTO t VALUES (1,100), (2, DEFAULT), (3)

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java [hunk 2]
```diff
@@ -69,6 +83,98 @@ public IgniteSqlToRelConvertor(
...
+        datasetStack.push(call);
+
+        RelNode rel = super.convertInsert(call);
+
+        datasetStack.pop();
+
+        return rel;
+    }
+
+    private static class DefaultChecker extends SqlShuttle {
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java:96]:
```suggestion
    private static class DefaultChecker extends SqlShuttle {
         boolean hasDefaults(SqlCall call) {
             try {
                 values.accept(checker);
                 return false;
             } catch (ControlFlowException e) {
                 return true;
             }
         }
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java (hunk 1 above)
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java:153]: Such kind of query format is not possible )

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java [hunk 3]
```diff
@@ -69,6 +83,101 @@ public IgniteSqlToRelConvertor(
...
+
+        for (SqlNode rowConstructor : values.getOperandList()) {
+            SqlCall rowConstructor0 = (SqlCall) rowConstructor;
+
+            List<Pair<RexNode, String>> exps = new ArrayList<>(targetFields.size());
+
+            int pos = 0;
+            int rowPos = 0;
+            for (RelDataTypeField fld : tblFields) {
+                if (fld.getName().equals(targetFields.get(rowPos))) {
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java:153]: I tried the following 
```
CREATE TABLE integers(i INTEGER PRIMARY KEY, j INTEGER DEFAULT 200, k INTEGER DEFAULT 100)
INSERT INTO integers (i, j) VALUES (1, DEFAULT)
```
and got the following exception
```
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 2 out of bounds for length 2
	at com.google.common.collect.RegularImmutableList.get(RegularImmutableList.java:77)
	at org.apache.calcite.util.Util$TransformingList.get(Util.java:2794)
	at org.apache.ignite.internal.sql.engine.prepare.IgniteSqlToRelConvertor.convertValuesImplEx(IgniteSqlToRelConvertor.java:153)
	at org.apache.ignite.internal.sql.engine.prepare.IgniteSqlToRelConvertor.convertValues(IgniteSqlToRelConvertor.java:125)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQueryRecursive(SqlToRelConverter.java:3781)
	at org.apache.ignite.internal.sql.engine.prepare.IgniteSqlToRelConvertor.convertQueryRecursive(IgniteSqlToRelConvertor.java:82)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertInsert(SqlToRelConverter.java:3825)
	at org.apache.ignite.internal.sql.engine.prepare.IgniteSqlToRelConvertor.convertInsert(IgniteSqlToRelConvertor.java:89)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQueryRecursive(SqlToRelConverter.java:3767)
	at org.apache.ignite.internal.sql.engine.prepare.IgniteSqlToRelConvertor.convertQueryRecursive(IgniteSqlToRelConvertor.java:82)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQuery(SqlToRelConverter.java:600)
	at org.apache.ignite.internal.sql.engine.prepare.IgnitePlanner.rel(IgnitePlanner.java:252)
	at org.apache.ignite.internal.sql.engine.prepare.PlannerHelper.optimize(PlannerHelper.java:83)
	at org.apache.ignite.internal.sql.engine.prepare.PrepareServiceImpl.lambda$prepareDml$4(PrepareServiceImpl.java:295)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	... 3 more
```

Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java:153]: @xtern good catch ! thanks !
fixed

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDmlTest.java [hunk 4]
```diff
@@ -575,40 +575,46 @@ private DefaultValueArg(String sqlType, String sqlVal, Object expectedVal) {
...
-                .returns(1, 2)
-                .returns(2, 3)
-                .returns(3, 2)
-                .returns(4, 4)
-                .returns(5, 2)
+        sql("CREATE TABLE integers(i INTEGER PRIMARY KEY, col1 INTEGER DEFAULT 200, col2 INTEGER DEFAULT 300)");
+
+        sql("INSERT INTO integers (i) VALUES (0)");
+        sql("INSERT INTO integers VALUES (1, DEFAULT, DEFAULT)");
+        sql("INSERT INTO integers(i, col2) VALUES (2, DEFAULT), (3, 4), (4, DEFAULT)");
```
Comment (xtern) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDmlTest.java:582]: I tried to swap fields order  "i, col2" => "col2, i"
```
sql("INSERT INTO integers(col2, i) VALUES (DEFAULT, 2), (4, 3), (DEFAULT, 4)");
```
and got the following exception
```
Caused by: java.lang.AssertionError: Field ordinal 1 is invalid for  type 'RecordType(INTEGER EXPR$2)'
	at org.apache.calcite.rex.RexBuilder.makeFieldAccess(RexBuilder.java:215)
	at org.apache.calcite.sql2rel.SqlToRelConverter.createInsertBlackboard(SqlToRelConverter.java:4065)
	at org.apache.calcite.sql2rel.SqlToRelConverter.collectInsertTargets(SqlToRelConverter.java:4139)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertColumnList(SqlToRelConverter.java:3989)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertInsert(SqlToRelConverter.java:3826)
	at org.apache.ignite.internal.sql.engine.prepare.IgniteSqlToRelConvertor.convertInsert(IgniteSqlToRelConvertor.java:89)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQueryRecursive(SqlToRelConverter.java:3767)
	at org.apache.ignite.internal.sql.engine.prepare.IgniteSqlToRelConvertor.convertQueryRecursive(IgniteSqlToRelConvertor.java:82)
	at org.apache.calcite.sql2rel.SqlToRelConverter.convertQuery(SqlToRelConverter.java:600)
	at org.apache.ignite.internal.sql.engine.prepare.IgnitePlanner.rel(IgnitePlanner.java:252)
	at org.apache.ignite.internal.sql.engine.prepare.PlannerHelper.optimize(PlannerHelper.java:83)
	at org.apache.ignite.internal.sql.engine.prepare.PrepareServiceImpl.lambda$prepareDml$4(PrepareServiceImpl.java:295)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	... 3 more
```

Comment (zstan) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDmlTest.java:582]: cool ! fixed

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java [hunk 5]
```diff
@@ -69,6 +83,111 @@ public IgniteSqlToRelConvertor(
...
+                if (operand.getKind() == SqlKind.DEFAULT) {
+                    RexNode def = ignTable.descriptor().newColumnDefaultValue(targetTable, mapping[pos], bb);
+
+                    exps.add(Pair.of(def, SqlValidatorUtil.alias(operand, pos)));
+                } else {
+                    exps.add(Pair.of(bb.convertExpression(operand), SqlValidatorUtil.alias(operand, pos)));
+                }
+
+                ++pos;
+            }
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java:178]: it seems to me that we can avoid last cycle.
Using something like this:
```suggestion
            for (String fld : targetFields) {
                for (int tblPos = 0; tblPos < tblFields.size(); tblPos++) {
                    if (!tblFields.get(tblPos).getName().equals(fld)) {
                        continue;
                    }

                    SqlNode operand = rowConstructor0.getOperandList().get(pos);

                    if (operand.getKind() == SqlKind.DEFAULT) {
                        RexNode def = ignTable.descriptor().newColumnDefaultValue(targetTable, tblPos, bb);

                        exps.add(Pair.of(def, SqlValidatorUtil.alias(operand, pos)));
                    } else {
                        exps.add(Pair.of(bb.convertExpression(operand), SqlValidatorUtil.alias(operand, pos)));
                    }

                    ++pos;

                    break;
                }
            }
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java [hunk 6]
```diff
@@ -69,6 +83,111 @@ public IgniteSqlToRelConvertor(
...
+        }
+
+        for (SqlNode rowConstructor : values.getOperandList()) {
+            SqlCall rowConstructor0 = (SqlCall) rowConstructor;
+
+            List<Pair<RexNode, String>> exps = new ArrayList<>(targetFields.size());
+
+            pos = 0;
+            for (String fld : targetFields) {
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java:166]: PMD check failed: Avoid unused local variables such as 'fld'

I suggest to replace it with a loop with a counter.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java [hunk 7]
```diff
@@ -69,6 +83,111 @@ public IgniteSqlToRelConvertor(
...
+                if (operand.getKind() == SqlKind.DEFAULT) {
+                    RexNode def = ignTable.descriptor().newColumnDefaultValue(targetTable, mapping[pos], bb);
+
+                    exps.add(Pair.of(def, SqlValidatorUtil.alias(operand, pos)));
+                } else {
+                    exps.add(Pair.of(bb.convertExpression(operand), SqlValidatorUtil.alias(operand, pos)));
+                }
+
+                ++pos;
+            }
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlToRelConvertor.java:178]: PMD check failed: Avoid unused local variables such as 'fld'

```suggestion
           for (pos = 0; pos < targetFields.size(); pos++) {
                SqlNode operand = rowConstructor0.getOperandList().get(pos);

                if (operand.getKind() == SqlKind.DEFAULT) {
                    RexNode def = ignTable.descriptor().newColumnDefaultValue(targetTable, mapping[pos], bb);

                    exps.add(Pair.of(def, SqlValidatorUtil.alias(operand, pos)));
                } else {
                    exps.add(Pair.of(bb.convertExpression(operand), SqlValidatorUtil.alias(operand, pos)));
                }
            }
```

**Pull request merged**